### ISSUE TITLE
consider indexing

### ISSUE THREAD

**Issue by maddanio** [niosus/EasyClangComplete#439]
Hi,

I just want to throw this in there. I would like to at least have a small discussion.

I love ecc, mostly for its aot compilation (i.e. telling me i messed up before running the build), completion, and showing of types. But I really really miss "go to implementation", and this requires indexing, since, unless inline, it cannot be inferred from the current translation unit alone. I now ended up using both ecc and rtags, but this is wasteful, as i now trigger ecc and clang whenever i save an edit.

I wonder how far away it would be for ecc to do indexing? Doesn't it already cache single TU indices anyway?

I would be willing to help, though to be honest i am already quite deep in work. but if that is what it takes i would see what i can do :).

**Comment by niosus** [comment]
Hey, thanks for reopening this discussion. 

I have been considering this for a long time now, but I don't really see any way to do it with the tools at hand. As far as I understand this, we will need to find all possible targets in the project and find out on the fly which TU does the current file belong to. I don't really see a feasible way of doing this currently. 

Currently, the TUs get cached, but completely independently. So a TU that relies on the files that are part of another TU does not know about it. And then the code just gets compiled, it is not linked in any way, so clang only cares about function declaration.

As discussed in issue #124 I still see no way of actually implementing it. Consider a simple use case. The code has 2 libraries `lib_a` and `lib_b` while editing files belonging to `lib_a` and using header from `lib_b` there is no way we can get function definitions if we do not link them. This is a whole new world of pain to even consider linking within this project.

And then again, as you pointed out, time. I have less and less time and will be writing my thesis in the next couple of months, so the time becomes a really valuable resource. I am being realistic in saying that I won't be able to invest enough time even if we find a good strategy to target this.

So, to sum up, I really hate to disappoint you, but I don't see any way of doing it now, but I am open to suggestions should you have any. I would love to see this happening in theory, but I don't think it can happen in my spare time. 

What I am considering is to add the results from sublime's internal ctags based index into the info message. So the function declaration will be guessed from semantics, while the definition - from index. I really think this is as good as it will get for this project. :)

**Comment by niosus** [comment]
Yeah, I understand that it is not the best solution, but definitely the best for now.
